scripts/dcmd.1
.TH DCMD 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
dcmd \- expand file lists of .dsc/.changes files in the command line
.SH SYNOPSIS
\fBdcmd\fR [\fIoptions\fR] [\fIcommand\fR] [\fIchanges-file\fR|\fIdsc-file\fR] ...
.SH DESCRIPTION
\fBdcmd\fR replaces any reference to a \fI.dsc\fR or \fI.changes\fR file in the
command line with the list of files in its 'Files' section, plus the
file itself.  It allows easy manipulation of all the files involved in
an upload (for \fI.changes\fR files) or a source package (for \fI.dsc\fR files).

If \fIcommand\fR is omitted (that is the first argument is an existing \fI.dsc\fR
or \fI.changes\fR file), the expanded list of files is printed to stdout, one file
by line. Useful for usage in backticks.
.SH OPTIONS
There are a number of options which may be used in order to select only a
subset of the files listed in the \fI.dsc\fR or \fI.changes\fR file. If a requested file
is not found, an error message will be printed.
.TP 14
.B \-\-dsc
Select the \fI.dsc\fR file.
.TP
.B \-\-schanges
Select \fI.changes\fR files for the 'source' architecture.
.TP
.B \-\-bchanges
Select \fI.changes\fR files for binary architectures.
.TP
.B \-\-changes
Select \fI.changes\fR files. Implies \fB\-\-schanges\fR and \fB\-\-bchanges\fR.
.TP
.B \-\-archdeb
Select architecture-dependent binary packages (\fI.deb\fR files).
.TP
.B \-\-indepdeb
Select architecture-independent binary packages (\fI.deb\fR files).
.TP
.B \-\-deb
Select binary packages (\fI.deb\fR files). Implies \fB\-\-archdeb\fR and \fB\-\-indepdeb\fR.
.TP
.B \-\-archudeb
Select architecture-dependent \fI.udeb\fR binary packages.
.TP
.B \-\-indepudeb
Select architecture-independent \fI.udeb\fR binary packages.
.TP
.B \-\-udeb
Select \fI.udeb\fR binary packages. Implies \fB\-\-archudeb\fR and \fB\-\-indepudeb\fR.
.TP
.BR \-\-tar ,\  \-\-orig
Select the upstream \fI.tar\fR file.
.TP
.BR \-\-diff ,\  \-\-debtar
Select the Debian \fI.debian.tar\fR or \fI.diff\fR file.
.PP
Each option may be prefixed by \fB\-\-no\fR to indicate that all files
\fInot\fR matching the specification should be selected.
.PP
It is not possible to combine positive filtering options (e.g. \fB\-\-dsc\fR)
and negative filtering options (e.g. \fB\-\-no\-changes\fR) in the same
\fBdcmd\fR invocation.
.TP
.B \-\-no\-fail\-on\-missing\fR, \fB\-r
If any of the requested files were not found, do not output an error.
.TP
.B \-\-package\fR, \fB\-p
Output package name part only.
.TP
.B \-\-sort\fR, \fB\-s
Sort output alphabetically.
.TP
.B \-\-tac\fR, \fB\-t
Reverse output order.

.SH "EXAMPLES"
Copy the result of a build to another machine:

.nf
$ dcmd scp rcs_5.7-23_amd64.changes elegiac:/tmp
rcs_5.7-23.dsc                  100%  490     0.5KB/s   00:00
rcs_5.7-23.diff.gz              100%   12KB  11.7KB/s   00:00
rcs_5.7-23_amd64.deb            100%  363KB 362.7KB/s   00:00
rcs_5.7-23_amd64.changes        100% 1095     1.1KB/s   00:00
$

$ dcmd \-\-diff \-\-deb scp rcs_5.7-23_amd64.changes elegiac:/tmp
rcs_5.7-23.diff.gz              100%   12KB  11.7KB/s   00:00
rcs_5.7-23_amd64.deb            100%  363KB 362.7KB/s   00:00
$
.fi

Check the contents of a source package:

.nf
$ dcmd md5sum rcs_5.7-23.dsc
8fd09ea9654cda128f8d5c337d3b8de7  rcs_5.7.orig.tar.gz
f0ceeae96603e823eacba6721a30b5c7  rcs_5.7-23.diff.gz
5241db1e231b1f43ae5514b63d2523f8  rcs_5.7-23.dsc
$

$ dcmd \-\-no\-diff md5sum rcs_5.7-23.dsc
8fd09ea9654cda128f8d5c337d3b8de7  rcs_5.7.orig.tar.gz
5241db1e231b1f43ae5514b63d2523f8  rcs_5.7-23.dsc
$
.fi

.SH "SEE ALSO"
.BR dpkg-genchanges (1),
.BR dpkg-source (1)
.SH AUTHOR
This program was written by Romain Francoise <rfrancoise@debian.org> and
is released under the GPL, version 2 or later.


debian/NEWS
devscripts (2.14.0) unstable; urgency=medium

  uscan now looks for upstream's keyring under debian/upstream/ as either
  signing-key.pgp or, for an armored keyring, signing-key.asc.  The previously
  used location, debian/upstream-signing-key.pgp, will be checked as a last
  resort for a transition period.

 -- James McCoy <jamessan@debian.org>  Sat, 25 Jan 2014 10:19:02 -0500

devscripts (2.11.9) unstable; urgency=low

  The default for mk-build-deps --tool option/MKBUILDDEPS_TOOL configuration
  changed from "apt-get" to "apt-get --no-install-recommends".

 -- James McCoy <jamessan@debian.org>  Sun, 17 Jun 2012 22:56:26 +0200

devscripts (2.11.7) unstable; urgency=low

  The heuristic used to determine what action to perform when no options are
  given to debchange now defaults to 'changelog'.  To stay with the previous
  behavior, DEBCHANGE_RELEASE_HEURISTIC=log can be added to the configuration
  file.

  Enabling the changelog heuristic also enables DEBCHANGE_MAINTTRAILER.  This
  can be disabled by adding DEBCHANGE_MAINTTRAILER=no to the configuration
  file.

 -- James McCoy <jamessan@debian.org>  Sun, 29 Apr 2012 18:38:46 -0400

devscripts (2.11.2) unstable; urgency=low

  Support for pgp-style commands has been removed from debsign/debrsign.  As
  such, the DEBSIGN_SIGNLIKE configuration variable has also been removed.

 -- James McCoy <jamessan@debian.org>  Tue, 15 Nov 2011 21:36:40 -0500

devscripts (2.10.54) unstable; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/dcmd.sh
#!/bin/sh
#
# dcmd: expand file lists of .dsc/.changes files in the command line
#
# Copyright (C) 2008 Romain Francoise <rfrancoise@debian.org>
# Copyright (C) 2008 Christoph Berg <myon@debian.org>
# Copyright (C) 2008 Adam D. Barratt <adsb@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

# Usage:
#
# dcmd replaces any reference to a .dsc or .changes file in the command
# line with the list of files in its 'Files' section, plus the
# .dsc/.changes file itself.
#
# $ dcmd sha1sum rcs_5.7-23_amd64.changes
# f61254e2b61e483c0de2fc163321399bbbeb43f1  rcs_5.7-23.dsc
# 7a2b283b4c505d8272a756b230486a9232376771  rcs_5.7-23.diff.gz
# e3bac970a57a6b0b41c28c615f2919c931a6cb68  rcs_5.7-23_amd64.deb
# c531310b18773d943249cfaa8b539a9b6e14b8f4  rcs_5.7-23_amd64.changes
# $

PROGNAME=`basename $0`

version () {
    echo \
"This is $PROGNAME, from the Debian devscripts package, version ###VERSION###
This code is copyright 2008 by Romain Francoise, all rights reserved.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later."
}

usage()
{
    printf "Usage: %s [options] [command] [dsc or changes file] [...]\n" $PROGNAME
}

endswith()
{
    case $1 in
	*$2) return 0 ;;
	*) return 1;;
    esac
}

# Instead of parsing the file completely as the previous Python
# implementation did (using python-debian), let's just select lines
# that look like they might be part of the file list.
RE="^ [0-9a-f]{32} [0-9]+ ((([a-zA-Z0-9_.-]+/)?[a-zA-Z0-9_.-]+|-) ([a-zA-Z]+|-) )?(.*)$"

maybe_expand()
{
    local dir
    local sedre
    if [ -e "$1" ] && (endswith "$1" .changes || endswith "$1" .dsc); then
	# Need to escape whatever separator is being used in sed expression so
	# it doesn't prematurely end the s command
	dir=$(dirname "$1" | sed 's/,/\\,/g')
	if [ "$(echo "$1" | cut -b1-2)" != "./" ]; then
	    sedre="\."
	fi
	sed --regexp-extended -n "s,$RE,$dir/\5,p" <"$1" | sed "s,^$sedre/,,"
    fi
}

DSC=1; BCHANGES=1; SCHANGES=1; ARCHDEB=1; INDEPDEB=1; TARBALL=1; DIFF=1
CHANGES=1; DEB=1; ARCHUDEB=1; INDEPUDEB=1; UDEB=1;
FILTERED=0; FAIL_MISSING=1

while [ $# -gt 0 ]; do
    TYPE=""
    case "$1" in
	--version|-v) version; exit 0;;
	--help|-h) usage; exit 0;;
	--no-fail-on-missing|-r) FAIL_MISSING=0;;
	--fail-on-missing) FAIL_MISSING=1;;
	--) shift; break;;
	--no-*)
	    TYPE=${1#--no-}
	    case "$FILTERED" in
		1)  echo "$PROGNAME: Can't combine --foo and --no-foo options" >&2;
		    exit 1;;
		0)  FILTERED=-1;;
	    esac;;
	--**)
	    TYPE=${1#--}
	    case "$FILTERED" in
		-1) echo "$PROGNAME: Can't combine --foo and --no-foo options" >&2;
		    exit 1;;
		0)  FILTERED=1; DSC=0; BCHANGES=0; SCHANGES=0; CHANGES=0
		    ARCHDEB=0; INDEPDEB=0; DEB=0; ARCHUDEB=0; INDEPUDEB=0
		    UDEB=0; TARBALL=0; DIFF=0;;
	    esac;;
	*) break;;
    esac

    case "$TYPE" in
	"") ;;
	dsc) [ "$FILTERED" = "1" ] && DSC=1 || DSC=0;;
	changes) [ "$FILTERED" = "1" ] &&
	    { BCHANGES=1; SCHANGES=1; CHANGES=1; } ||
	    { BCHANGES=0; SCHANGES=0; CHANGES=0; } ;;
	bchanges) [ "$FILTERED" = "1" ] && BCHANGES=1 || BCHANGES=0;;
	schanges) [ "$FILTERED" = "1" ] && SCHANGES=1 || SCHANGES=1;;
	deb) [ "$FILTERED" = "1" ] &&
	    { ARCHDEB=1; INDEPDEB=1; DEB=1; } ||
	    { ARCHDEB=0; INDEPDEB=0; DEB=0; };;
	archdeb) [ "$FILTERED" = "1" ] && ARCHDEB=1 || ARCHDEB=0;;
	indepdeb) [ "$FILTERED" = "1" ] && INDEPDEB=1 || INDEPDEB=0;;
	udeb) [ "$FILTERED" = "1" ] &&
	    { ARCHUDEB=1; INDEPUDEB=1; UDEB=1; } ||
	    { ARCHUDEB=0; INDEPUDEB=0; UDEB=0; };;
	archudeb) [ "$FILTERED" = "1" ] && ARCHUDEB=1 || ARCHUDEB=0;;
	indepudeb) [ "$FILTERED" = "1" ] && INDEPUDEB=1 || INDEPUDEB=0;;
	tar|orig) [ "$FILTERED" = "1" ] && TARBALL=1 || TARBALL=0;;
	diff) [ "$FILTERED" = "1" ] && DIFF=1 || DIFF=0;;
	*) echo "$PROGNAME: Unknown option '$1'" >&2; exit 1;;
    esac
    shift
done

cmd=
args=""
while [ $# -gt 0 ]; do
    arg="$1"
    shift
    temparg="$(maybe_expand "$arg")"
    if [ -z "$temparg" ]; then
	if [ -z "$cmd" ]; then
	    cmd="$arg"
	    continue
	fi
	# Not expanded, so simply add to argument list
	args="$args
$arg"
    else
	SEEN_INDEPDEB=0; SEEN_ARCHDEB=0; SEEN_SCHANGES=0; SEEN_BCHANGES=0
	SEEN_INDEPUDEB=0; SEEN_ARCHUDEB=0; SEEN_UDEB=0;
	SEEN_TARBALL=0; SEEN_DIFF=0; SEEN_DSC=0
	MISSING=0
	newarg=""
	# Output those items from the expanded list which were
	# requested, and record which files are contained in the list
	eval "$(echo "$temparg" | while read THISARG; do
	    if [ -z "$THISARG" ]; then
		# Skip
		:
	    elif endswith "$THISARG" _all.deb; then
		[ "$INDEPDEB" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_INDEPDEB=1;"
	    elif endswith "$THISARG" .deb; then
		[ "$ARCHDEB" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_ARCHDEB=1;"
	    elif endswith "$THISARG" _all.udeb; then
		[ "$INDEPUDEB" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_INDEPUDEB=1;"
	    elif endswith "$THISARG" .udeb; then
		[ "$ARCHUDEB" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_ARCHUDEB=1;"
	    elif endswith "$THISARG" .tar.gz || \
		 endswith "$THISARG" .tar.xz || \
		 endswith "$THISARG" .tar.lzma || \
		 endswith "$THISARG" .tar.bz2; then
		[ "$TARBALL" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_TARBALL=1;"
	    elif endswith "$THISARG" _source.changes; then
		[ "$SCHANGES" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_SCHANGES=1;"
	    elif endswith "$THISARG" .changes; then
		[ "$BCHANGES" = "0" ] || echo "newarg\"\$newarg
$THISARG\";"
		echo "SEEN_BCHANGES=1;"
	    elif endswith "$THISARG" .dsc; then
		[ "$DSC" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_DSC=1;"
	    elif endswith "$THISARG" .diff.gz; then
		[ "$DIFF" = "0" ] || echo "newarg=\"\$newarg
$THISARG\";"
		echo "SEEN_DIFF=1;"
	    elif [ "$FILTERED" != "1" ]; then
		# What is it? Output anyway
		echo "newarg=\"\$newarg
$THISARG\";"
	    fi
	done)"

	INCLUDEARG=1
	if endswith "$arg" _source.changes; then
	    [ "$SCHANGES" = "1" ] || INCLUDEARG=0
	    SEEN_SCHANGES=1
	elif endswith "$arg" .changes; then
	    [ "$BCHANGES" = "1" ] || INCLUDEARG=0
	    SEEN_BCHANGES=1
	elif endswith "$arg" .dsc; then
	    [ "$DSC" = "1" ] || INCLUDEARG=0
	    SEEN_DSC=1
	fi

	if [ "$FAIL_MISSING" = "1" ] && [ "$FILTERED" = "1" ]; then
	    if [ "$CHANGES" = "1" ]; then
		if [ "$SEEN_SCHANGES" = "0" ] && [ "$SEEN_BCHANGES" = "0" ]; then
		    MISSING=1; echo "$arg: .changes fiie not found" >&2
		fi
	    else
		if [ "$SCHANGES" = "1" ] && [ "$SEEN_SCHANGES" = "0" ]; then
		    MISSING=1; echo "$arg: source .changes file not found" >&2
		fi
		if [ "$BCHANGES" = "1" ] && [ "$SEEN_BCHANGES" = "0" ]; then
		    MISSING=1; echo "$arg: binary .changes file not found" >&2
		fi
	    fi

	    if [ "$DEB" = "1" ]; then
		if  [ "$SEEN_INDEPDEB" = "0" ] && [ "$SEEN_ARCHDEB" = "0" ]; then
		    MISSING=1; echo "$arg: binary packages not found" >&2
		fi
	    else
		if [ "$INDEPDEB" = "1" ] && [ "$SEEN_INDEPDEB" = "0" ]; then
		    MISSING=1; echo "$arg: arch-indep packages not found" >&2
		fi
		if [ "$ARCHDEB" = "1" ] && [ "$SEEN_ARCHDEB" = "0" ]; then
		    MISSING=1; echo "$arg: arch-dep packages not found" >&2
		fi
	    fi

	    if [ "$UDEB" = "1" ]; then
		if [ "$SEEN_INDEPUDEB" = "0" ] && [ "$SEEN_ARCHUDEB" = "0" ]; then
		    MISSING=1; echo "$arg: udeb packages not found" >&2
		fi
	    else
		if [ "$INDEPUDEB" = "1" ] && [ "$SEEN_INDEPUDEB" = "0" ]; then
		    MISSING=1; echo "$arg: arch-indep udeb packages not found" >&2
		fi
		if [ "$ARCHUDEB" = "1" ] && [ "$SEEN_ARCHUDEB" = "0" ]; then
		    MISSING=1; echo "$arg: arch-dep udeb packages not found" >&2
		fi

	    fi

	    if [ "$DSC" = "1" ] && [ "$SEEN_DSC" = "0" ]; then
		MISSING=1; echo "$arg: .dsc file not found" >&2
	    fi
	    if [ "$TARBALL" = "1" ] && [ "$SEEN_TARBALL" = "0" ]; then
		MISSING=1; echo "$arg: upstream tar not found" >&2
	    fi
	    if [ "$DIFF" = "1" ] && [ "$SEEN_DIFF" = "0" ]; then
		MISSING=1; echo "$arg: Debian diff not found" >&2
	    fi

	    [ "$MISSING" = "0" ] || exit 1
	fi

	args="$args
$newarg"
	[ "$INCLUDEARG" = "0" ] || args="$args
$arg"
    fi
done

IFS='
'
if [ -z "$cmd" ]; then
    for arg in $args; do
	echo $arg
    done
    exit 0
fi

exec $cmd $args
